src/aspire/classification/averager2d.py

```diff
@@ -384,44 +404,46 @@ def align(self, classes, reflections, basis_coefficients):
 
         n_classes = classes.shape[0]
 
-        # Compute the shifts. Roll array so 0 is first.
-        x_shifts = np.roll(
-            np.arange(-self.n_x_shifts, self.n_x_shifts + 1), -self.n_x_shifts
-        )
-        y_shifts = np.roll(
-            np.arange(-self.n_y_shifts, self.n_y_shifts + 1), -self.n_y_shifts
+        # Create a search grid and force initial pair to (0,0)
+        # This is done primarily in case of a tie later, we would take unshifted.
+        x_shifts, y_shifts = self._shift_search_grid(
+            self.src.L, self.radius, roll_zero=True
         )
-        # Above rolls should force initial pair of shifts to (0,0).
-        # This is done primarily in case of a tie later we would take unshifted.
-        assert (x_shifts[0], y_shifts[0]) == (0, 0)
 
         # These arrays will incrementally store our best alignment.
         rotations = np.empty(classes.shape, dtype=self.dtype)
         correlations = np.ones(classes.shape, dtype=self.dtype) * -np.inf
         shifts = np.empty((*classes.shape, 2), dtype=int)
 
-        if basis_coefficients is None:
-            # Retrieve image coefficients, this is bad, it load all images.
-            # TODO: Refactor this s.t. the following code blocks and super().align
-            #   only require coefficients relating to their class.  See _cls_images.
-            basis_coefficients = self.composite_basis.evaluate_t(self.src.images[:])
-
         # We want to maintain the original coefs for the base images,
         #  because we will mutate them with shifts in the loop.
-        original_coef = basis_coefficients[classes[:, 0], :]
+        if basis_coefficients is None:
+            original_coef = self.composite_basis.evaluate_t(
+                self._cls_images(classes[:, 0], src=self.src)
+            )
+        else:
+            original_coef = basis_coefficients[classes[:, 0], :].copy()
+
+        # Sanity check the original_coef shape
         assert original_coef.shape == (n_classes, self.alignment_basis.count)
 
         # Loop over shift search space, updating best result
-        for x, y in product(x_shifts, y_shifts):
+        for x, y in zip(x_shifts, y_shifts):
             shift = np.array([x, y], dtype=int)
             logger.debug(f"Computing rotational alignment after shift ({x},{y}).")
 
             # Shift the coef representing the first (base) entry in each class
             #   by the negation of the shift
             # Shifting one image is more efficient than shifting every neighbor
-            basis_coefficients[classes[:, 0], :] = self.alignment_basis.shift(
-                original_coef, -shift
-            )
+            if basis_coefficients is not None:
+                basis_coefficients[classes[:, 0], :] = self.alignment_basis.shift(
+                    original_coef, -shift
+                )
+            else:
+                # This variable will be checked and apply by super class (BFR.align)
```

Comment:
Something like "Store latest shift so that super class can undo/reset it. This allows us to ... "?

Reply:
Sure, changed, slightly different wording.